### PR TITLE
feat(): Azure provider and checks

### PR DESCRIPTION
**Pull request by n4ch04**
### Context 

Adds azure provider with checks from the following services:

- IAM
- Defender
- Storage

### Description

Adds azure provider with checks from the following services:

- IAM
- Defender
- Storage

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: providers/azure/azure_provider.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+from msgraph.core import GraphClient
+
+from lib.logger import logger
+from providers.azure.lib.audit_info.audit_info import azure_audit_info
+from providers.azure.lib.audit_info.models import Azure_Identity_Info
+
+
+class Azure_Provider:
+    def __init__(self):
+        logger.info("Instantiating azure provider ...")
```
Comment (jfagoagas) [inline providers/azure/azure_provider.py]:
```suggestion
        logger.info("Instantiating Azure Provider ...")
```

File: providers/azure/azure_provider.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    try:
+        logger.info("Azure provider: validating service principal credentials ...")
+        result = client.get("/servicePrincipals/").json()
+        if "value" in result:
+            for sp in result["value"]:
+                if sp["appId"] == azure_identity.app_id:
+                    azure_identity.id = sp["id"]
+    except Exception as error:
+        logger.critical(f"{error.__class__.__name__} -- {error}")
```
Comment (jfagoagas) [inline providers/azure/azure_provider.py]:
```suggestion
        logger.critical(f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}")
```

File: providers/azure/azure_provider.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        sys.exit()
+    else:
+        return azure_identity
+
+
+def azure_provider_set_session():
+    logger.info("Setting azure session ...")
```
Comment (jfagoagas) [inline providers/azure/azure_provider.py]:
```suggestion
    logger.info("Setting Azure session ...")
```

File: providers/azure/azure_provider.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+        domain_result = client.get("/domains").json()
+        if "value" in domain_result:
+            if "id" in domain_result["value"][0]:
+                azure_audit_info.audited_account = domain_result["value"][0]["id"]
+        subscriptions_client = SubscriptionClient(
+            credential=azure_audit_info.credentials
+        )
+        for subscription in subscriptions_client.subscriptions.list():
+            logger.info(subscription.__dict__)
```
Comment (jfagoagas) [inline providers/azure/azure_provider.py]: Is this intended or only for debugging purposes?

File: providers/azure/azure_provider.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            azure_audit_info.subscriptions.update(
+                {subscription.display_name: subscription.subscription_id}
+            )
+    except Exception as error:
+        logger.critical(f"{error.__class__.__name__} -- {error}")
```
Comment (jfagoagas) [inline providers/azure/azure_provider.py]:
```suggestion
        logger.critical(f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}] -- {error}")
```

File: providers/azure/services/defender/defender_ensure_defender_for_app_services_is_on/defender_ensure_defender_for_app_services_is_on.metadata.json [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_app_services_is_on/defender_ensure_defender_for_app_services_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_arm_is_on/defender_ensure_defender_for_arm_is_on.metadata.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_arm_is_on/defender_ensure_defender_for_arm_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_azure_sql_databases_is_on/defender_ensure_defender_for_azure_sql_databases_is_on.metadata.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_azure_sql_databases_is_on/defender_ensure_defender_for_azure_sql_databases_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_containers_is_on/defender_ensure_defender_for_containers_is_on.metadata.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_containers_is_on/defender_ensure_defender_for_containers_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_databases_is_on/defender_ensure_defender_for_databases_is_on.metadata.json [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_databases_is_on/defender_ensure_defender_for_databases_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_dns_is_on/defender_ensure_defender_for_dns_is_on.metadata.json [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  "Remediation": {
+    "Code": {
+      "CLI": "",
+      "NativeIaC": "",
+      "Other": "",
+      "Terraform": ""
+    },
+    "Recommendation": {
+      "Text": "By default, Microsoft Defender for Cloud is not enabled for your App Service instances. Enabling the Defender security service for App Service instances allows for advanced security defense using threat detection capabilities provided by Microsoft Security Response Center.",
+      "Url": "Enable Microsoft Defender for DNS"
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_dns_is_on/defender_ensure_defender_for_dns_is_on.metadata.json]:
```suggestion
      "Url": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_dns_is_on/defender_ensure_defender_for_dns_is_on.metadata.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_dns_is_on/defender_ensure_defender_for_dns_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_keyvault_is_on/defender_ensure_defender_for_keyvault_is_on.metadata.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_keyvault_is_on/defender_ensure_defender_for_keyvault_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_os_relational_databases_is_on/defender_ensure_defender_for_os_relational_databases_is_on.metadata.json [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_os_relational_databases_is_on/defender_ensure_defender_for_os_relational_databases_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_server_is_on/defender_ensure_defender_for_server_is_on.metadata.json [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_server_is_on/defender_ensure_defender_for_server_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_sql_servers_is_on/defender_ensure_defender_for_sql_servers_is_on.metadata.json [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_sql_servers_is_on/defender_ensure_defender_for_sql_servers_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/defender/defender_ensure_defender_for_storage_is_on/defender_ensure_defender_for_storage_is_on.metadata.json [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_storage_is_on/defender_ensure_defender_for_storage_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/iam/iam_subscription_roles_owner_custom_not_created/iam_subscription_roles_owner_custom_not_created.metadata.json [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/iam/iam_subscription_roles_owner_custom_not_created/iam_subscription_roles_owner_custom_not_created.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/iam/iam_subscription_roles_owner_custom_not_created/iam_subscription_roles_owner_custom_not_created.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+                report.region = iam_client.region
+                report.status = "PASS"
+                report.status_extended = f"Role {role.name} from subscription {subscription} is not a custom owner role"
+                for scope in role.assignable_scopes:
+                    if search("^/.*", scope):
+                        for permission_item in role.permissions:
+                            for action in permission_item.actions:
+                                if action == "*":
+                                    report.status = "FAIL"
+                                    report.status_extended = f"Role {role.name} from subscription {subscription} is a custom owner role"
```
Comment (jfagoagas) [inline providers/azure/services/iam/iam_subscription_roles_owner_custom_not_created/iam_subscription_roles_owner_custom_not_created.py:22]: I think this status is not quite accurate, isn't it?

File: providers/azure/services/storage/storage_blob_public_access_level_is_disabled/storage_blob_public_access_level_is_disabled.metadata.json [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_blob_public_access_level_is_disabled/storage_blob_public_access_level_is_disabled.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/storage/storage_default_network_access_rule_is_denied/storage_default_network_access_rule_is_denied.metadata.json [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_default_network_access_rule_is_denied/storage_default_network_access_rule_is_denied.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/storage/storage_ensure_azure_services_are_trusted_to_access_is_enabled/storage_ensure_azure_services_are_trusted_to_access_is_enabled.metadata.json [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_ensure_azure_services_are_trusted_to_access_is_enabled/storage_ensure_azure_services_are_trusted_to_access_is_enabled.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/storage/storage_ensure_encryption_with_customer_managed_keys/storage_ensure_encryption_with_customer_managed_keys.metadata.json [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_ensure_encryption_with_customer_managed_keys/storage_ensure_encryption_with_customer_managed_keys.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/storage/storage_ensure_minimum_tls_version_12/storage_ensure_minimum_tls_version_12.metadata.json [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_ensure_minimum_tls_version_12/storage_ensure_minimum_tls_version_12.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/storage/storage_infrastructure_encryption_is_enabled/storage_infrastructure_encryption_is_enabled.metadata.json [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_infrastructure_encryption_is_enabled/storage_infrastructure_encryption_is_enabled.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/storage/storage_secure_transfer_required_is_enabled/storage_secure_transfer_required_is_enabled.metadata.json [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_secure_transfer_required_is_enabled/storage_secure_transfer_required_is_enabled.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/services/storage/storage_service.py [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+                    )
+        except Exception as error:
+            logger.error(
+                f"{error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
+            )
+        else:
+            return storage_accounts
+
+
+# def __get_storage_account_keys_rotation_time__(self):
```
Comment (jfagoagas) [inline providers/azure/services/storage/storage_service.py]:
```suggestion
```

**Review by jfagoagas — APPROVED** [review]

File: providers/azure/services/defender/defender_ensure_defender_for_cosmosdb_is_on/defender_ensure_defender_for_cosmosdb_is_on.metadata.json [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "Categories": [],
+  "Tags": {
+    "Tag1Key": "value",
+    "Tag2Key": "value"
+  },
+  "DependsOn": [],
+  "RelatedTo": [],
+  "Notes": "",
+  "Compliance": []
```
Comment (jfagoagas) [inline providers/azure/services/defender/defender_ensure_defender_for_cosmosdb_is_on/defender_ensure_defender_for_cosmosdb_is_on.metadata.json]:
```suggestion
  "Notes": ""
```

File: providers/azure/azure_provider.py (hunk 4 above)
Comment (n4ch04) [reply providers/azure/azure_provider.py]:
```suggestion
        
```

File: providers/azure/services/iam/iam_subscription_roles_owner_custom_not_created/iam_subscription_roles_owner_custom_not_created.py (hunk 19 above)
Comment (n4ch04) [reply providers/azure/services/iam/iam_subscription_roles_owner_custom_not_created/iam_subscription_roles_owner_custom_not_created.py:22]: The check is looking for roles in a subscription with same permissions of built in role owner, to the message is accurate

**Pull request merged**